File: Day1/bank.cpp

#include <bits/stdc++.h>
using namespace std;

// Base Loan class (Abstraction)
class Loan {
protected:
    double principal;   
    double rate;        
    int years;          
public:
    Loan(double p, double r, int y) : principal(p), rate(r), years(y) {}
    virtual double calculateInterest() const = 0;
    virtual ~Loan() = default;
};

// Car Loan (Open for extension)
class CarLoan : public Loan {
public:
    CarLoan(double p, double r, int y) : Loan(p, r, y) {}

    double calculateInterest() const override {
        return (principal * rate * years) / 100.0;
    }
};

// Home Loan (Open for extension)
class HomeLoan : public Loan {
public:
    HomeLoan(double p, double r, int y) : Loan(p, r, y) {}

    double calculateInterest() const override {
        double amount = principal * pow((1 + rate / 100.0), years);
        return amount - principal;
    }
};

// Bank Service (DIP - works with abstraction)
class BankService {
private:
    Loan* loan;  
public:
    BankService(Loan* l) : loan(l) {}

    void showInterest() {
        cout << "Interest Payable: " << loan->calculateInterest() << endl;
    }
};


int main() {
    CarLoan car(500000, 8, 5);   
    HomeLoan home(2000000, 7, 10); 
    BankService carLoanService(&car);
    BankService homeLoanService(&home);
    cout <<"Car Loan";
    carLoanService.showInterest();

    cout <<"Home Loan";
    homeLoanService.showInterest();

    return 0;
}

File: Day 01/bank.java
import java.lang.Math;

// Base Loan class (Abstraction)
abstract class Loan {
    protected double principal;
    protected double rate;
    protected int years;

    public Loan(double principal, double rate, int years) {
        this.principal = principal;
        this.rate = rate;
        this.years = years;
    }

    // Abstract method
    public abstract double calculateInterest();
}

// Car Loan (Open for extension)
class CarLoan extends Loan {
    public CarLoan(double principal, double rate, int years) {
        super(principal, rate, years);
    }

    @Override
    public double calculateInterest() {
        return (principal * rate * years) / 100.0;
    }
}

// Home Loan (Open for extension)
class HomeLoan extends Loan {
    public HomeLoan(double principal, double rate, int years) {
        super(principal, rate, years);
    }

    @Override
    public double calculateInterest() {
        double amount = principal * Math.pow((1 + rate / 100.0), years);
        return amount - principal;
    }
}

// Bank Service (DIP - works with abstraction)
class BankService {
    private Loan loan;

    public BankService(Loan loan) {
        this.loan = loan;
    }

    public void showInterest() {
        System.out.println("Interest Payable: " + loan.calculateInterest());
    }
}

// Main class
public class Main {
    public static void main(String[] args) {
        CarLoan car = new CarLoan(500000, 8, 5);
        HomeLoan home = new HomeLoan(2000000, 7, 10);

        BankService carLoanService = new BankService(car);
        BankService homeLoanService = new BankService(home);

        System.out.println("Car Loan");
        carLoanService.showInterest();

        System.out.println("Home Loan");
        homeLoanService.showInterest();
    }
}
